yellow-paper/docs/contracts/index.md

```diff
@@ -12,6 +13,138 @@ The purpose of the L1 contracts are simple:
+        self.last_block_ts = block.timestamp
+        self.block_number = 0
+
+    def process(
+        self,
+        header: Header, 
+        archive: Fr, 
+        txs_hash: Fr, 
+        l1_to_l2_msgs: Fr[], 
+        l2_to_l1_msgs: Fr[], 
+        proof: Proof
+    ):
+        assert self.body_available(header, txs_hash, l1_to_l2_msgs, l2_to_l1_msgs)
+        assert self.validate_header(header)
+        assert self.verifier.verify(proof, header, archive)
+        assert self.registry.inbox.consume(l1_to_l2_msgs)
+        assert self.registry.outbox.insert(l2_to_l1_msgs)
+        self.archive = archive
+
+    def body_available(
+        self,
+        content_hash: Fr, 
+        txs_hash: Fr, 
+        l1_to_l2_msgs: Fr[], 
+        l2_to_l1_msgs: Fr[]
+    ) -> bool:
+        assert self.registry.availability_oracle.is_available(txs_hash)
+        in_hash = SHA256(pad(l1_to_l2_msgs))
+        out_hash = MerkleTree(pad(l2_to_l1_msgs), SHA256)
+        return content_hash == SHA256(txs_hash, out_hash, in_hash)
+    )
+
+    def validate_header(
+        self,
+        header: Header
+    ) -> bool:
+        assert header.global_variables.block_number = self.block_number + 1
+        assert header.global_variables.chain_id == block.chain_id
+        assert header.global_variables.version == self.version 
+        assert header.global_variables.timestamp < block.timestamp
+        assert header.global_variables.timestamp > self.last_block_ts
+        assert header.archive == self.archive
+
+        return True
+```
+
+
+### Availability Oracle
+The state transitioner should be connected to an oracle which addresses the availability condition. 
+
+For the case of a rollup, this "oracle" will be deriving the `TxsHash` from calldata and blobs. For a validium it should be connected to a bridge that it can use to verify that the data is available on the other chain. 
+
+For a generic DA that publishes data commitments to Ethereum, the oracle could be a snark proof that opens the data commitment from the bridge and computes the `TxsHash` from it.
+
+By having the availability oracle be independent from state progression we can even do multi-transaction blocks, e.g., use multiple transactions or commitments from other DA layers to construct the `TxsHash` for a large block. 
+
+
+### Registry
+To keep one location where all the core rollup contracts can be found, we have a registry contract. The registry is a contract that holds the current and historical addresses of the core rollup contracts. The addresses of a rollup deployment are contained in a snapshot, and the registry is tracking version-snapshot pairs. Depending on the upgrade scheme, it might be used to handle upgrades, or it could entirely be removed. It is generally the one address that a node MUST know about, as it can then tell the node where to find the remainder of the contracts. This is for example used when looking for the address new L2 blocks should be published to.
```

Comment:
Where are all the "core rollup contracts" defined? Is it this page? Perhaps we could list all of the L1 smart contracts at the top of the page, with links to the sections which then explain each of those contracts?